File: linked_lists/lc430.js
/**
 * // Definition for a Node.
 * function Node(val,prev,next,child) {
 *    this.val = val;
 *    this.prev = prev;
 *    this.next = next;
 *    this.child = child;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var flatten = function(head) {
    if (head === null) {
        return head;
    }

    let flat = function(subhead) {
        let pt = subhead;
        let lastNode = pt; // Store the last processed node

        while (pt !== null) {
            if (pt.child !== null) {
                let nextNode = pt.next;
                let lastChild = flat(pt.child);
                pt.next = pt.child;
                pt.child.prev = pt;
                pt.child = null;
                lastChild.next = nextNode;

                if (nextNode !== null) {
                    nextNode.prev = lastChild;
                }

                lastNode = lastChild; // Update last processed node
            }

            lastNode = pt; // Update last processed node even if there's no child
            pt = pt.next;
        }

        return lastNode; // Return the last processed node
    };

    flat(head);

    return head;
};

File: linked_lists/lc430.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return head
        
        def flat(subhead):
            #keep a pointer on the current subhead
            pt = subhead
            #keep a pointer on what's the new tail
            lastNode = pt

            #enter main loop to explore depth
            while pt is not None:
                #if you see there is a child, explore and reassign
                if pt.child is not None:
                    #keep track of the current nextNode
                    nextNode = pt.next
                    lastChild = flat(pt.child) #look for the lastNode of that child

                    #cleanup
                    pt.next = pt.child
                    pt.child.prev = pt
                    pt.child = None
                    lastChild.next = nextNode
                    
                    if nextNode is not None:
                        nextNode.prev = lastChild

                #otherwise scoot along
                #record current potential last node
                lastNode = pt
                pt = pt.next
            
            return lastNode
        
        flat(head)
        return head
